New Histogram field mapper that supports percentiles aggregations.

iverase opened the pull request:
This PR explores the addition of a new histogram field mapper that consist in a pre-aggregated format of numerical data to be used in percentiles aggregations.

### Mapper

The new field is defined in a mapping using the following structure:

```
PUT /example
{
    "mappings": {
        "properties": {
            "aggregated": {
                "type": "histogram"
            }
        }
    }
}
```

And it can be populated using the following structure:

```
POST /example/_doc
{
    "aggregated" : {
        "values" :[0.1, 0.2, 0.3, 0.4, 0.5],
        "counts" : [5, 3, 14, 6, 4]
    }
}
```

where `values` is an array of doubles and `counts` is an array of integers and must have the same length. This format is up for discussion but the reasons to choose this format is that they can be easily generated from the existing histograms.

For TDigest, they can be generated using the following code (see [example](https://github.com/iverase/elasticsearch/blob/c4bfdb773a13fb29df96df9c683dbcd1e21e4583/x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/mapper/HistogramAggregationTests.java#L136)):

```
                List<Double> values = new ArrayList<>();
                List<Integer> counts = new ArrayList<>();
                Collection<Centroid> centroids = histogram.centroids();
                for (Centroid centroid : centroids) {
                    values.add(centroid.mean());
                    counts.add(centroid.count());
                }
```

For HDR histograms, hey can be generated using the following code (see [example](https://github.com/iverase/elasticsearch/blob/c4bfdb773a13fb29df96df9c683dbcd1e21e4583/x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/mapper/HistogramAggregationTests.java#L41)):

```
                List<Double> values = new ArrayList<>();
                List<Integer> counts = new ArrayList<>();
                Iterator<DoubleHistogramIterationValue> iterator = histogram.recordedValues().iterator();
                while (iterator.hasNext()) {
                    DoubleHistogramIterationValue histValue = iterator.next();
                    values.add(histValue.getValueIteratedTo());
                    counts.add(Math.toIntExact(histValue.getCountAtValueIteratedTo()));
                }
```

This structure is stored as a binary doc value

### Aggregations

This field can be used in standard `percentile` and `percentile_ranks` aggregations. They can be used together with standard numeric fields. In order for this aggregations to support this new format the following interfaces has been created:

- `IndexHistogramFieldData`: A new specialisation of field data for histograms.
- `AtomicHistogramFieldData`: A new specialisation of atomic field data for histograms.
- `HistogramValues`: The docVales returned by the atomic field data.
- `HistogramValue`: A doc value representing one of those histograms.

The aggregations has been updated accordingly to be able to understand this new field data.


### Open questions

- Accuracy and suitability of the format.

- The field mapper does not support as input an array of histograms, does it need it to support it?

- Overflow of aggregations? As we are adding pre-aggregated data, it can theoretically happen that aggregation histogram can more easily overflow (I think the maximum number of total counts is LONG.MAX_VALUE).


Relates #48578

iverase committed:
Add HistogramField.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Aggregations)

iverase committed:
checkStyle

iverase committed:
more checkStyle

jpountz commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractHDRPercentilesAggregator.java:
```diff
@@ -106,6 +118,36 @@ public void collect(int doc, long bucket) throws IOException {
+                    // initialisation and doesn't need resizing.
+                    state.setAutoResize(true);
+                    states.set(bucket, state);
+                }
```
can we extract the logic of creating a new histogram to a separate method that is used by both collectNumeric and collectHistogramValues?

iverase replied:
done

jpountz commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/HDRPercentilesAggregatorFactory.java:
```diff
@@ -57,23 +56,23 @@
+                                          Map<String, Object> metaData) throws IOException {
         return new HDRPercentilesAggregator(name, valuesSource, searchContext, parent, percents, numberOfSignificantValueDigits, keyed,
-                config.format(), pipelineAggregators, metaData);
+            config.format(), pipelineAggregators, metaData);
```
can you try to undo formatting changes in this file and a couple other ones to keep the diff readable?

jpountz commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/PercentileRanksAggregationBuilder.java:
```diff
@@ -80,7 +79,7 @@
         PARSER = new ConstructingObjectParser<>(PercentileRanksAggregationBuilder.NAME, false,
             (a, context) -> new PercentileRanksAggregationBuilder(context, (List) a[0]));
-        ValuesSourceParserHelper.declareNumericFields(PARSER, true, false, false);
+        ValuesSourceParserHelper.declareAnyFields(PARSER, true, true);
```
Can we try to be less lenient and require a field that is either numeric or a histogram?

iverase replied:
Yes, that makes sense and that is the tricky part. @not-napoleon I might want to borrow a bit of your thoughts on how to do this in the best way possible.

not-napoleon replied:
The parser doesn't currently support that, and right now the convention is to use ANY for aggregations that can accept more than one type of input.  Cleaning that up is a goal of the ValuesSource refactor effort I'm currently working on (see #42949)

jpountz replied:
Thanks @not-napoleon, let's keep this TODO for later then @iverase ?

iverase replied:
Agreed. I think to be able to support that we require the effort that @not-napoleon is working on (thanks for taking a look!).

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,459 @@
+                                                    return getHistogramValue(values.binaryValue());
+                                                } catch (IOException e) {
+                                                    throw new IllegalStateException("Cannot load doc value", e);
+                                                }
```
Let's make `histogram()` throw an `IOException` for consistency with `BinaryDocValues#binaryValue`?

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,459 @@
+                                    } catch (IOException e) {
+                                        throw new IllegalStateException("Cannot load doc values", e);
+                                    }
+                                }
```
We should throw an UnsupportedOperationException in the two above methods.

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,459 @@
+                        @Override
+                        public SortField sortField(Object missingValue, MultiValueMode sortMode,
+                                                   XFieldComparatorSource.Nested nested, boolean reverse) {
+                            return null;
```
this should throw an UnsupportedOperationException too

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,459 @@
+                private HistogramValue getHistogramValue(final BytesRef bytesRef) throws IOException {
+                    final ByteBufferStreamInput streamInput = new ByteBufferStreamInput(
+                        ByteBuffer.wrap(bytesRef.bytes, bytesRef.offset, bytesRef.length));
+                    final int numValues = streamInput.readVInt();
```
We could also avoid storing the length and consider the iterator exhausted when all bytes have been read.

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,459 @@
+        context.path().add(simpleName());
+        try {
+            List<Double> values = null;
+            List<Integer> counts = null;
```
let's use native variants, e.g. IntArrayList and DoubleArrayList?

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,459 @@
+            if (values.size() == 0) {
+                throw new MapperParsingException("error parsing field ["
+                    + name() + "], arrays for values and counts cannot be empty");
+            }
```
I wonder whether we should actually fail it or not.

jpountz replied:
On the other hand, maybe we should require that values come in order and fail if there are duplicate values?

iverase replied:
Respect empty arrays, yes we can just ignore it instead of failing.

I like he idea of requiring values to be ordered and disallow duplicates. Before implementing I will wait for more input.

iverase replied:
I updated the code so in case of empty arrays we just ignore it as we do with null values.

In addition we require now that values are provided in increasing order.

jpountz reviewed (commented):
I like it a lot. My main concern is that the parsing of aggregations was made lenient for this to work by accepting any field in the parser. I think it's fine that we use generic objects internally like `ValuesSource` and do `instanceof` calls to see how these objects should be handled. However I'd like the parser to be strict and fail if a field is provided that is neither a numeric field or a histogram field?

Let's add docs?

> Accuracy and suitability of the format.

It looks good to me.

> The field mapper does not support as input an array of histograms, does it need it to support it?

I don't think it needs to. We could document that this field only supports single values, like we do for vector fields (you might want to borrow the logic that vectors used to make sure fields are actually single-valued).

> Overflow of aggregations? As we are adding pre-aggregated data, it can theoretically happen that aggregation histogram can more easily overflow (I think the maximum number of total counts is LONG.MAX_VALUE).

I think we should ignore this problem on our end and push this problem to the upstream libraries, since this is a general problem that they have given that they allow collecting multiple instances of the same value at once. The fact that you require counts to be integers and reject longs sounds to me like the right thing to do on our end.

iverase committed:
Addressed part of the review

iverase committed:
Extract the logic of creating a new histogram to a separate method

iverase committed:
Addressed more comments.

iverase committed:
formatting

felixbarny commented:
Really looking forward to this 😍 
The API also looks great to me!

Question: does the `values` array have to contain the same values for each doc? In other words, does the first doc determine the structure of the histogram?

Is it possible to extend the range in subsequent documents? Example:

```
POST /example/_doc
{
    "aggregated" : {
        "values" :[0.1, 0.2],
        "counts" : [1, 2]
    }
}

POST /example/_doc
{
    "aggregated" : {
        "values" :[0.1, 0.2, 0.3],
        "counts" : [1, 2, 3]
    }
}
```


Is it possible to use different buckets in subsequent documents? Example:
```
POST /example/_doc
{
    "aggregated" : {
        "values" :[0.1, 0.2],
        "counts" : [1, 2]
    }
}

POST /example/_doc
{
    "aggregated" : {
        "values" :[0.15, 0.2, 0.25],
        "counts" : [1, 2, 3]
    }
}
```
Is it possible to omit buckets with a value of zero? Example:
```
POST /example/_doc
{
    "aggregated" : {
        "values" :[0.1, 0.2, 0.3],
        "counts" : [1, 2, 3]
    }
}

POST /example/_doc
{
    "aggregated" : {
        "values" :[0.1, 0.3],
        "counts" : [1, 0, 3]
    }
}
```

These are not necessarily requirements we have in APM, just trying to get a feel for what's possible.

iverase committed:
extract logic for getting histogram in TDigest

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractTDigestPercentilesAggregator.java:
```diff
@@ -75,18 +77,22 @@ public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
         }
         final BigArrays bigArrays = context.bigArrays();
-        final SortedNumericDoubleValues values = valuesSource.doubleValues(ctx);
+        if (valuesSource instanceof ValuesSource.Histogram) {
```
Is it a good idea to be doing an `instanceof` operation for every document here? Maybe the performance of `instanceof` is negligible these days but if not, given this is run in a very tight loop, maybe we should be choosing a "mode" in the constructor and then checking that enum/boolean each time instead?

jpountz replied:
@colings86 I think this code is called once for every segment actually.

colings86 replied:
oh, you are right, ignore this then :)

colings86 commented on x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/mapper/HistogramAggregationTests.java:
```diff
@@ -0,0 +1,237 @@
+import java.util.List;
+
+
+public class HistogramAggregationTests extends ESSingleNodeTestCase {
```
This naming is a bit confusing since there is also a histogram aggregation which is completely different to this new field. Maybe we should call this `HistogramPercentileAggregationTests` or something else?

iverase replied:
Originally created this test to check different approaches. I renamed it following your suggestions.

iverase committed:
remove unused imports

iverase committed:
rename test class

iverase committed:
Detect in the constructor if we expect histogram value source
and use that to decide the collecting method

iverase committed:
revert last change

iverase commented:
@felixbarny 

> Is it possible to extend the range in subsequent documents? 
> Is it possible to use different buckets in subsequent documents?

Yes that is possible as every histogram is independent to the others so it can have different buckets and different number of buckets.

> Is it possible to omit buckets with a value of zero?

The count for a bucket can be zero but the bucket must be present. In general the length of the `values` and `counts` arrays must be the same. We are still considering some other changes, in particular we might require that buckets are ordered and you cannot have twice the same bucket.

colings86 commented:
> you cannot have twice the same bucket.

One thing to note is that it's possible in TDigest particularly to have multiple buckets with the same value. This can happen if the same value is inserted enough times to exceed the threshold for the centroid which will force a new centroid with the same value to be created. IF we move to reject histograms with multiple of the same bucket, it will make the code for extracting the TDigest histogram (from your PR description) less simple since the user will need to keep track of whether the value has changed when moving to the next centroid.

I don't see a problem with requiring ordered buckets though

jpountz commented:
Good point Colin, I agree that rejecting duplicates would make it tricky to work with t-digests.

iverase commented:
If we allow duplicates, I don't think we require buckets to be ordered. wdyt?

jpountz commented:
I think we should at least make sure values are sorted in the doc-value representation. Then we could either sort buckets ourselves or require users to provide us with sorted buckets.

iverase committed:
Values must be provided in increasing order

iverase committed:
Handling null value and do not fail if arrays are empty, trate it as a
null value

iverase committed:
Handle ignore malformed properly

iverase committed:
Merge branch 'master' into histogramField

iverase committed:
initial documentation for the new field

iverase committed:
initial documentation for the new field

jrodewig commented on docs/reference/aggregations/metrics/percentile-aggregation.asciidoc:
```diff
@@ -3,8 +3,8 @@
-can be extracted either from specific numeric fields in the documents, or
-be generated by a provided script.
+can be extracted either from specific numeric or histogram fields in the documents,
+or be generated by a provided script.
```
The comma is unneeded here. However, it may be more clear to swap the order of the phrases:

> These values can be generated by a provided script or extracted from
specific numeric or histogram fields in the documents.

jrodewig commented on docs/reference/aggregations/metrics/percentile-rank-aggregation.asciidoc:
```diff
@@ -3,8 +3,8 @@
-can be extracted either from specific numeric fields in the documents, or
-be generated by a provided script.
+can be extracted either from specific numeric fields or histogram in the documents,
+or be generated by a provided script.
```
Same comment here:

> These values can be generated by a provided script or extracted from
specific numeric or histogram fields in the documents.

jrodewig commented on docs/reference/mapping/types.asciidoc:
```diff
@@ -32,6 +32,7 @@ string::         <<text,`text`>> and <<keyword,`keyword`>>
 <<ip>>::            `ip` for IPv4 and IPv6 addresses
 <<completion-suggester,Completion datatype>>::
                     `completion` to provide auto-complete suggestions
+<<histogram>>:: Record pre-aggregated numerical values to be used in percentiles aggregations.
```
I'd include the literal to be consistent with the rest of the list:

```suggestion
<<histogram>>:: `histogram` for pre-aggregated numerical values for percentiles aggregations.
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+
+A `histogram` field stores pre-aggregated numerical data represented as a histogram. These histograms
+are defined by a set of pairs, one being the value representing the bucket and the count of that value.
+THe values should be defined accordingly to the type of aggregation this field will be used for.
```
```suggestion
The values should be defined accordingly to the type of aggregation this field will be used for.
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+and <<search-aggregations-metrics-percentile-rank-aggregation, percentile ranks>> aggregations are supported.
+
+Histograms are indexed by defining two arrays, one containing the `values` (or buckets) for the histogram and the other containing
+the `counts` for such values. Those two arrays must have the same length.
```
This section is great, but I think it's missing some crucial information:

- The `values` array stores `double` floats
- The `counts` array stores `long` integers
- The field data is stored as doc values and not indexed
- Because it's not indexed, it can't be used for most queries

I also think some of this section can get repetitive and needs some chunking. I left a suggestion below, but don't feel obligated to use it.

```
A  field to store pre-aggregated numerical data representing a histogram.
This data is defined using two paired arrays:

* A `values` array of <<number,`double`>> numbers, representing the buckets for
  the histogram. These values must be provided in ascending order.
* A corresponding `counts` array of  <<number,`long`>> numbers, representing how
  many values fall into each bucket.

Because the elements in the `values` array correspond to the elements in the
same position of the `count` array, these two arrays must have the same length.

[[histogram-uses]]
==== Uses

`histogram` fields are primarily intended for use with aggregations. To make it
more readily accessible for aggregations, `histogram` field data is stored as a
binary <<doc-values,doc values>> and not indexed.

Because the data is not indexed, you only can use `histogram` fields for the
following aggregations and queries:

* <<search-aggregations-metrics-percentile-aggregation,percentiles>> aggregation
* <<search-aggregations-metrics-percentile-rank-aggregation,percentile ranks>> aggregation
* <<query-dsl-exists-query,exists>> query

We recommend you define the buckets in the `values` array based on the type of aggregation you intended to use.
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+[[mapping-types-histogram-building-histogram]]
+==== Building a histogram
+
+When using a histogram as part of an aggregation, the accuracy of the results will depend on how the
```
```suggestion
When using a histogram as part of an aggregation, the accuracy of the results depends on how the
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+==== Building a histogram
+
+When using a histogram as part of an aggregation, the accuracy of the results will depend on how the
+histogram was constructed. It is important to consider the percentiles aggregation mode that will be used
```
```suggestion
histogram was constructed. It is important to consider the percentiles aggregation mode that used
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+
+When using a histogram as part of an aggregation, the accuracy of the results will depend on how the
+histogram was constructed. It is important to consider the percentiles aggregation mode that will be used
+in order to build it. Some possibilities are:
```
```suggestion
to build it. Some possibilities include:
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+can be built by using the mean value of the centroids and the centroid's count. If the algorithm has already
+started to approximate the percentiles, this inaccuracy is carried over in the histogram.
+
+- For the https://github.com/HdrHistogram/HdrHistogram[HDR Histogram] (High Dynamic Range Histogram) method, histograms
```
It may be better to link to our HDR docs here: `<<_hdr_histogram,HDR>> (High Dynamic Range)...`
```suggestion
- For the https://github.com/HdrHistogram/HdrHistogram[HDR] (High Dynamic Range) histogram method, histograms
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+
+Internally, each document's histogram is encoded as a binary
+doc value. Its size in bytes is at most `12 * numValues`,
+where `numValues` is the length of the provided arrays.
```
It may be helpful to move this information up higher.

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+arrays are treated as null.
+
+NOTE: `rank_feature` fields do not support querying or sorting. They may
+only be used within <<query-dsl-exists-query, exists>> queries.
```
This is pretty important information. I included this content in my earlier suggestion, but otherwise you may want to move it above the example and condense it into a single `[IMPORTANT]` admonition.

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+NOTE: `histogram` fields with null field are ignored. Fields with empty `values` and `counts`
+arrays are treated as null.
+
+NOTE: `rank_feature` fields do not support querying or sorting. They may
```
```suggestion
NOTE: `histogram` fields do not support querying or sorting. They may
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+--------------------------------------------------
+
+<1> Values for each bucket. Values in the array are treated as doubles and must be given in
+increasing order. In case if TDDigest histograms this value represents the mean value.
```
```suggestion
increasing order. For<<search-aggregations-metrics-percentile-aggregation-approximation,T-Digest>> histograms, this value represents the mean value.
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+
+<1> Values for each bucket. Values in the array are treated as doubles and must be given in
+increasing order. In case if TDDigest histograms this value represents the mean value.
+In case of HDR histograms this represents the value iterated to.
```
```suggestion
For <<_hdr_histogram,High Dynamic Range (HDR)>> histograms, this represents the iterated value.
```

jrodewig commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,87 @@
+
+Histograms are indexed by defining two arrays, one containing the `values` (or buckets) for the histogram and the other containing
+the `counts` for such values. Those two arrays must have the same length.
+
```
Not strictly needed for MVP, but it might be nice to add some text for context and split up the index creation and document indexing snippets.

For example:

```
[[histogram-ex]]
==== Examples

The following <<indices-create-index, create index>> API request creates a new index with two field mappings:

* `my_histogram`, a `histogram` field used to store percentile data
* `my_text`, a `keyword` field used to store a title for the histogram

[ INSERT CREATE INDEX SNIPPET ]
...

The following <<docs-index_,index>> API requests store pre-aggregated for two histograms: `histogram_`` and `histogram_2`.

[ INSERT DOC INDEX SNIPPET ]
...
```

Providing an example use case for the data may also be helpful. For example, the histograms could represent load time, similar to the percentile aggs docs. Not required for MVP though.

iverase committed:
Addressed docs review

iverase committed:
Add HistogramFieldTypeTests

jtibshirani commented:
This looks like an exciting new field type! I had a few questions around the API:
* It seems like the `histogram` type is intended for both HDR histograms and t-digest? I wonder if a name like `sketch` or `data_sketch` would be more accurate, since t-digest doesn’t actually produce histograms, it gives a weighted list of cluster centroids. We’ve also brainstormed support for [moment sketches](http://www.bailis.org/papers/moments-vldb2018.pdf) which takes a very different approach and stores a list of moments of the data.
* If the field contains the output of a t-digest sketch, then to correctly compute `percentiles` it seems the user must be sure to use the `tdigest` mode. What happens if they accidentally use `hdr` mode, will it run without error but the results are slightly off? Do they need to supply the same `compression` that was used to originally create the t-digest sketch? (I am quite new to t-digest, apologies if these questions don’t make sense.)
* The histogram is represented as two fields `values` and `counts`. This differs from the way we represent histograms in some other places, like the [`histogram` aggregation](https://www.elastic.co/guide/en/elasticsearch/reference/7.4/search-aggregations-bucket-histogram-aggregation.html). Parallel arrays can also be harder to work with than a map or list of pairs (although this seems like less of a concern as they’re only used as an input parameter). The current API has some nice advantages, just wanted to point out some other considerations to make sure we’re happy with the decision.

jpountz reviewed (commented):
I left a couple comments but it looks almost ready to me. I wonder whether we should de-emphasize the fact that it is a pre-aggregated percentiles sketch from the docs. Percentiles are our motivation for adding this field, but I can imagine us adding support for sums or averages (weighted by counts) to this field in the future? (This is a non-important side thought that I had while reading docs, feel free to ignore for now.)

jpountz commented:
Regarding @jtibshirani 's point about whether this field is best implemented via parallel arrays (`{ 
values: [5, 7], counts: [8, 2] }`) or a list of entries (`[ { key: 5, count: 7 }, { key: 8, count: 2 } ]`) I wonder whether @felixbarny has an opinion?

iverase committed:
address last review comments

iverase commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,462 @@
+            }
+            if (values.size() == 0) {
+                context.path().remove();
+                return;
```
I agree, this check has been removed and a binaryDocValue of zero length will be generated.

iverase commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,462 @@
+                    if (counts.get(i) < 0) {
+                        throw new MapperParsingException("error parsing field ["
+                            + name() + "], ["+ COUNTS_FIELD + "] elements must be >= 0 but got " + counts.get(i));
+                    }
```
make sense, added code to skip zero `counts`

iverase committed:
Merge branch 'master' into histogramField

iverase commented:
@elasticmachine test elasticsearch-ci/2

iverase commented:
@elasticmachine please run elasticsearch-ci/2

felixbarny commented:
I like the compactness of the parallel arrays.

exekias commented:
One question, as I'm not sure I fully understand it from the signature. Does this work with cumulative counters? Meaning that count values are forever growing for subsequent reports, and store the total amount of occurrences in the bucket (and not the ones from the last sample).

We find this very useful when storing metrics, as it helps with data loss, for instance. This also what we see in other vendors (like Prometheus/OpenMetrics). Behind the scenes they just calculate the rate first and do percentile calculations afterwards.

iverase committed:
Merge branch 'master' into histogramField

# Conflicts:
#	x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/AnalyticsPlugin.java

iverase committed:
Merge branch 'master' into histogramField

# Conflicts:
#	server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceConfig.java
#	server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceType.java

iverase committed:
Make sure that in ignore malformed we move to the end of the
field

iverase commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,462 @@
+            if (ignoreMalformed.value() == false) {
+                throw new MapperParsingException("failed to parse field [{}] of type [{}]",
+                    ex, fieldType().name(), fieldType().typeName());
+            }
```
I added my own logic to skip the field to the end. Not sure if there is some other utility to what I am doing here by hand.

jpountz replied:
This is what XContentSubParser has been designed for. See https://github.com/elastic/elasticsearch/pull/35603. Maybe it would be more robust? By the way looking at the latest version of GeoShapeFieldMapper, it looks like it no longer handles ignoreMalformed correctly, or am I misreading it cc @imotov ?

iverase commented:
@jpountz , @polyfractal I have addressed the latest comments. I added code to make sure that we skip the whole field in case the ignoreMalformed option is set and there is an error.

I had a look into the current docs and I am not sure how to improve, it looks fine to me but any advise is welcome!

iverase commented:
@elasticmachine run elasticsearch-ci/default-distro

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,482 @@
+                                                                   Map<String, Object> node, ParserContext parserContext)
+                throws MapperParsingException {
+            Builder builder = new HistogramFieldMapper.Builder(name);
+            parseField(builder, name, node, parserContext);
```
Let's not call `parseField`, none of the properties that it parses are supported by this field?

iverase replied:
I am supporting them:

* user can disable docValues
* If user set index, index options or stored values, an error will be thrown.

I am not handling boost, similarity and copy_to. Shall we throw an error if user defines those fields?

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -0,0 +1,482 @@
+                    } else if (count > 0) {
+                        // we do not add elements with count == 0
+                        streamOutput.writeDouble(values.get(i));
+                        streamOutput.writeVInt(count);
```
I'd suggest putting the count before the values, it might make it easier to better compress in the future by stealing bits of the count.

jpountz reviewed (commented):
I'd like the ignoreMalformed logic to be a bit more robust, but other than that it looks good to me.

polyfractal commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,116 @@
+histogram was constructed. It is important to consider the percentiles aggregation mode that will be used
+to build it. Some possibilities include:
+
+- For the <<search-aggregations-metrics-percentile-aggregation, T-Digest>> mode, histograms
```
Hmm, trying to tweak this a little to make it more explicit, so the user knows what the value/count fields do.

> - For the <<search-aggregations-metrics-percentile-aggregation, T-Digest>> mode, the `values` array represents the mean centroid positions and the `counts` array represents the number of values that are attributed to each centroid. If the algorithm has already started to approximate the percentiles, this inaccuracy is carried over in the histogram.

WDYT?

polyfractal commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,116 @@
+can be built by using the mean value of the centroids and the centroid's count. If the algorithm has already
+started to approximate the percentiles, this inaccuracy is carried over in the histogram.
+
+- For the <<_hdr_histogram,High Dynamic Range (HDR)>> histogram mode, histograms
```
Similarly,

> - For the <<_hdr_histogram,High Dynamic Range (HDR)>> histogram mode, the `values` array represents fixed upper limits of each bucket interval, and the `counts` array represents the number of values that are attributed to each interval. This implementation maintains a fixed worse-case percentage error (specified as a number of significant digits), therefore the value used when generating the histogram would be the maximum accuracy you can achieve at aggregation time.

??

polyfractal commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,116 @@
+can be created by using the recorded values and the count at that value. This implementation maintains a fixed worse-case
+percentage error (specified as a number of significant digits), therefore the value used when generating the histogram
+would be the maximum accuracy you can achieve at aggregation time.
+
```
Perhaps another sentence/paragraph at the end?

> The `histogram` field is "algorithm agnostic" and does not store data specific to either T-Digest or HDRHistogram.  While this means the field can technically be aggregated with either algorithm, in practice the user should chose one algorithm and index data in that manner (e.g. centroids for T-Digest or intervals for HDRHistogram) to ensure best accuracy.

Or something similar... trying to convey to the user that how they index the data is important and they should chose upfront.

polyfractal reviewed (commented):
I think the docs look good. I just left a few minor comments/tweaks to make a certain aspect more explicit, otherwise I think they are fine :+1: 

Been following the code changes from afar, think Adrien has the code review aspect covered so I'll defer to him there :)

iverase committed:
address review comments

iverase commented:
@jpountz, I changed the logic so now I am using XContentSubParser to handle ignore malformed (as it should be).

My only doubt is about parse fields. I am supporting doc values so a user can disable them. It might be useful if a user wants to store the histogram fields (already parsed) without doc values so at a later stage it can reindex them again with them, wdyt?

should be handle boost, similarity and copy_to and throw an error if the user defines them?

jpountz commented:
Given that it's always easier to add features than remove them, I'd be in favor of only supporting `ignore_malformed` for now?

iverase committed:
remove support for parsed fields

iverase commented:
Great, that makes things simpler. I removed support for parse fields

jpountz commented on x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java:
```diff
@@ -451,27 +451,16 @@ public void parse(ParseContext context) throws IOException {
+            if (subParser != null) {
+                while (token != null) {
+                    token = subParser.nextToken();
                 }
```
Do `subParser.close()` instead?

jpountz commented on docs/reference/mapping/types/histogram.asciidoc:
```diff
@@ -0,0 +1,119 @@
+`histogram` fields are primarily intended for use with aggregations. To make it
+more readily accessible for aggregations, `histogram` field data is stored as a
+binary <<doc-values,doc values>> and not indexed. Its size in bytes is at most
+`12 * numValues`, where `numValues` is the length of the provided arrays.
```
I think it's actually 13 since vints can take up to 5 bytes.

iverase commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

iverase commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into histogramField

iverase committed:
addressed last comments

iverase committed:
Merge branch 'histogramField' of github.com:iverase/elasticsearch into histogramField
